### PR TITLE
make fsautocomplete working without change user config

### PR DESCRIPTION
**Pull request by s-kostyaev**
There is no need to force user update their configuration if we already know where to find executable.

@razzmatazz 

Relates #3350

### CODE REVIEW

**Comment by razzmatazz** [comment]
Hmm this will not work on windows and actually depends on where dotnet_home env points to..

**Comment by razzmatazz** [comment]
Again this is more generally a queation to @yyoncho on how we deal with broken/unconfigured  exec-path in lsp mode? What is the policy?

**Comment by s-kostyaev** [comment]
> dotnet_home

How it looks like on windows? I can add more branches...

**Comment by razzmatazz** [comment]
> > dotnet_home
> 
> How it looks like on windows? I can add more branches...

There is some info here:
- https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-global-tool

but to reiterate, I think we should check what other lsp clients impl do with broken exec-path before merging. I agree with the intent but I am a bit wary for adding too much magic if other clients (node npm?) fix this some other way in lsp mode..

**Comment by s-kostyaev** [comment]
@razzmatazz like this?

**Comment by s-kostyaev** [comment]
> Again this is more generally a question to @yyoncho on how do we deal with broken/unconfigured exec-path in lsp mode? What is the policy?

example:
```elisp
(defcustom lsp-clients-flow-server "flow"
  "The Flow executable to use.
Leave as just the executable name to use the default behavior of
finding the executable with variable `exec-path'."
  :group 'lsp-flow
  :risky t
  :type 'file)
```

**Comment by razzmatazz** [comment]
I don’t know. A good approach maybe would be to indicate the possible fix to the user somehow once we cannot find fsautocomplete on exec-path? But when and how?

There is lsp-doctor I think.., but that is not easily discoverable..

**Comment by s-kostyaev** [comment]
@yyoncho we need you :smile:

**Comment by razzmatazz** [comment]
@jcs090218 too :)

**Comment by s-kostyaev** [comment]
If we decide to merge it - it will work with automatically installed `fsautocomlete`. In other case we will install it, but it will not work without changing user configuration.

**Comment by yyoncho** [comment]
It looks fine to me. I read the `flow` example as an instruction.

**Comment by yyoncho** [comment]
IMHO if the user typically does not have this binary on the path and the case handled by this PR will be hit by all users if they follow the default installation steps then it makes sense to have the PR merged. lsp-mode should work as much OOTB as possible. If this piece of code improves the experience I don't see anything against it as long as it will be useful for the users.

**Comment by s-kostyaev** [comment]
> IMHO if the user typically does not have this binary on the path and the case handled by this PR will be hit by all users if they follow the default installation steps then it makes sense to have the PR merged. lsp-mode should work as much OOTB as possible. If this piece of code improves the experience I don't see anything against it as long as it will be useful for the users.

That's why I have created this PR. @razzmatazz @TOTBWF What do you think?

**Comment by TOTBWF** [comment]
I don't write a ton of F# these days so I'm not super equipped to answer this, but this seems reasonable enough to me. Anything that makes the initial onboarding experience simpler is an easy win!

**Comment by s-kostyaev** [comment]
@yyoncho can we merge it?

**Comment by razzmatazz** [comment]
> @yyoncho can we merge it?

I think the agreement is that this is a good approach after all, to minimize setup where possible. I will test the Linux code-path and then I can merge.

**Review by razzmatazz — APPROVED** [review]
I have tested the changes and it seems to work for me, on Linux at least (when simulating `(executable-find "fsac")` to be failing

**Comment by yyoncho** [comment]
> @yyoncho can we merge it?

@s-kostyaev 

I believe that decisions like that should be taken by fsharp/csharp maintainers/contributors/users. I mean, the decision is mainly affecting them.

**Pull request merged**
